src/Common/tests/System/Net/Http/Http2LoopbackServer.cs

```diff
@@ -61,623 +70,64 @@ private Http2LoopbackServer(Http2Options options)
-
-            return true;
-        }
-
-        public async Task<Frame> ReadFrameAsync(TimeSpan timeout)
-        {
-            using CancellationTokenSource timeoutCts = new CancellationTokenSource(timeout);
-            return await ReadFrameAsync(timeoutCts.Token);
-        }
-
-        private async Task<Frame> ReadFrameAsync(CancellationToken cancellationToken)
-        {
-            // Prep the timeout cancellation token.
-
-            // First read the frame headers, which should tell us how long the rest of the frame is.
-            byte[] headerBytes = new byte[Frame.FrameHeaderLength];
-
-            try
-            {
-                if (!await FillBufferAsync(headerBytes, cancellationToken).ConfigureAwait(false))
-                {
-                    return null;
-                }
-            }
-            catch (IOException)
-            {
-                // eat errors when client aborts connection and return null.
-                return null;
-            }
-
-            Frame header = Frame.ReadFrom(headerBytes);
-
-            // Read the data segment of the frame, if it is present.
-            byte[] data = new byte[header.Length];
-            if (header.Length > 0 && !await FillBufferAsync(data, cancellationToken).ConfigureAwait(false))
-            {
-                throw new Exception("Connection stream closed while attempting to read frame body.");
-            }
-
-            if (_ignoreSettingsAck && header.Type == FrameType.Settings && header.Flags == FrameFlags.Ack)
-            {
-                _ignoreSettingsAck = false;
-                return await ReadFrameAsync(cancellationToken).ConfigureAwait(false);
-            }
-
-            if (_ignoreWindowUpdates && header.Type == FrameType.WindowUpdate)
-            {
-                return await ReadFrameAsync(cancellationToken).ConfigureAwait(false);
-            }
-
-            // Construct the correct frame type and return it.
-            switch (header.Type)
-            {
-                case FrameType.Settings:
-                    return SettingsFrame.ReadFrom(header, data);
-                case FrameType.Data:
-                    return DataFrame.ReadFrom(header, data);
-                case FrameType.Headers:
-                    return HeadersFrame.ReadFrom(header, data);
-                case FrameType.Priority:
-                    return PriorityFrame.ReadFrom(header, data);
-                case FrameType.RstStream:
-                    return RstStreamFrame.ReadFrom(header, data);
-                case FrameType.Ping:
-                    return PingFrame.ReadFrom(header, data);
-                case FrameType.GoAway:
-                    return GoAwayFrame.ReadFrom(header, data);
-                default:
-                    return header;
-            }
-        }
-
-        // Reset and return underlying networking objects.
-        public (Socket, Stream) ResetNetwork()
+        private void RemoveInvalidConnections()
```

Comment:
How will this behave when on an architecture without cache coherency, if `IsValid` is `false` on another thread but `true` for this one?

Reply:
@scalablecory it shouldn't matter that much, indices are not used anywhere except for tracking if there is one vs multiple connections. If you have one then you probably won't use threads, if you have multi then it doesn't matter